forge-core/src/main/java/forge/util/StreamUtil.java

```diff
@@ -0,0 +1,109 @@
+
+public class StreamUtil {
+
+    private StreamUtil(){}
+
+    /**
+     * Reduces a stream to a random element of the stream. Used with {@link Stream#collect}.
+     * Result will be wrapped in an Optional, absent only if the stream is empty.
+     */
+    public static <T> Collector<T, ?, Optional<T>> random() {
+        return new RandomCollectorSingle<>();
+    }
+
+    /**
+     * Selects a number of items randomly from this stream. Used with {@link Stream#collect}.
+     * @param count Number of elements to select from the stream.
+     */
+    public static <T> Collector<T, ?, List<T>> random(int count) {
+        return new RandomCollectorMulti<>(count);
+    }
+
+    private static abstract class RandomCollector<T, O> implements Collector<T, RandomReservoir<T>, O> {
+        private final int size;
+        RandomCollector(int size) {
+            this.size = size;
+        }
+
+        @Override
+        public Supplier<RandomReservoir<T>> supplier() {
+            return () -> new RandomReservoir<>(size);
+        }
+
+        @Override
+        public BiConsumer<RandomReservoir<T>, T> accumulator() {
+            return RandomReservoir::accumulate;
+        }
+
+        @Override
+        public BinaryOperator<RandomReservoir<T>> combiner() {
+            return (first, second) -> {
+                //There's probably a way to adapt the Random Reservoir method
+                //so that two partially processed lists can be combined into one.
+                //But I have no idea what that is.
+                throw new UnsupportedOperationException("Parallel streams not supported.");
+            };
+        }
+
+        private final EnumSet<Characteristics> CHARACTERISTICS = EnumSet.of(Characteristics.UNORDERED);
+        @Override
+        public Set<Characteristics> characteristics() {
+            return CHARACTERISTICS;
+        }
+    }
+
+    private static class RandomCollectorSingle<T> extends RandomCollector<T, Optional<T>> {
+        RandomCollectorSingle() {
+            super(1);
+        }
+
+        @Override
+        public Function<RandomReservoir<T>, Optional<T>> finisher() {
+            return (chosen) -> chosen.samples.isEmpty() ? Optional.empty() : Optional.of(chosen.samples.get(0));
+        }
+    }
+
+    private static class RandomCollectorMulti<T> extends RandomCollector<T, List<T>> {
+        RandomCollectorMulti(int size) {
+            super(size);
+        }
+
+        @Override
+        public Function<RandomReservoir<T>, List<T>> finisher() {
+            return (chosen) -> chosen.samples;
+        }
+    }
+
+    private static class RandomReservoir<T> {
+        final int maxSize;
+        ArrayList<T> samples;
+        int sampleCount = 0;
+
+        public RandomReservoir(int size) {
+            this.maxSize = size;
+            this.samples = new ArrayList<>(size);
+        }
+
+        public void accumulate(T next) {
+            sampleCount++;
+            if(sampleCount < maxSize) {
```

Comment:
I believe you introduced a logic error here

Reply:
Good catch. I tried a few different implementations of that collector to try and get parallel support in before giving up and going back to something close to the original. Must have missed that `<=` in the process.